pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+//! # BlockRewards Pallet
+//!
+//! The BlockRewards pallet provides functionality for distributing rewards to
+//! different accounts with different currencies.
+//! The distribution happens when an session (a constant time interval) finalizes.
+//! Users cannot stake manually as their collator membership is syncronized via
+//! a provider.
+//! Thus, when new collators join, they will automatically be staked and vice-versa
+//! when collators leave, they are unstaked.
+//!
+//! The BlockRewards pallet provides functions for:
+//!
+//! - Claiming the reward given for a staked currency. The reward will be the native network's token.
+//! - Admin methods to configure sessions, currencies and reward rates as well as any user's stake.
+//!
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+pub mod migrations;
+pub mod weights;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub use cfg_traits::{
+	ops::{EnsureAdd, EnsureAddAssign},
+	rewards::{AccountRewards, CurrencyGroupChange, DistributedRewards, GroupRewards},
+};
+use cfg_types::tokens::CurrencyId as CfgCurrencyId;
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		fungibles::Mutate, tokens::Balance, Currency as CurrencyT, OnUnbalanced, OneSessionHandler,
+	},
+	DefaultNoBound,
+};
+pub use frame_support::{
+	storage::{bounded_btree_map::BoundedBTreeMap, transactional},
+	transactional,
+};
+use frame_system::pallet_prelude::*;
+use num_traits::sign::Unsigned;
+pub use pallet::*;
+pub use sp_runtime::Saturating;
+use sp_runtime::{traits::Zero, FixedPointOperand, SaturatedConversion};
+use sp_std::{mem, vec::Vec};
+use weights::WeightInfo;
+
+#[derive(
+	Encode, Decode, DefaultNoBound, Clone, TypeInfo, MaxEncodedLen, PartialEq, RuntimeDebugNoBound,
+)]
+#[scale_info(skip_type_params(T))]
```

Comment:
No, it has existed ever since `MaxEncodedLen` was added. It fixes compiler issues that arise from the derive macro for `T`. If we did not derive all types from `T::Config`, then deriving would be entirely fine.